r/tests/testthat/test-dplyr-funcs-datetime.R

```diff
@@ -118,6 +118,51 @@ test_that("errors in strptime", {
   )
 })
 
+test_that("strptime returns NA when format doesn't match the data", {
+  df <- tibble(
+    str_date = c("2022-02-07", "2012/02-07", "1975/01-02", "1981/01-07")
+  )
+
+  expect_equal(
+    df %>%
+      arrow_table() %>%
+      mutate(
+        r_obj_parsed_date = strptime("03-27/2022", format = "%m-%d/%Y"),
+        r_obj_parsed_na = strptime("03-27/2022", format = "Y%-%m-%d")) %>%
+      collect(),
+    tibble(
+      str_date = c("2022-02-07", "2012/02-07", "1975/01-02", "1981/01-07"),
+      r_obj_parsed_date = as.POSIXct(rep("2022-03-27", 4)),
+      r_obj_parsed_na = as.POSIXct(rep(NA, 4))
+    ),
+    ignore_attr = "tzone"
+  )
```

Comment:
Besides tzone, we expect dplyr and arrow execution methods to be the same now, yeah? Could we use `compare_dplyr_binding`?

Reply:
Nope, `base::strptime()` and the arrow binding return different types. `ignore_attr` cannot help here since that covers only classes, not types. `base::strptime()` returns a list (`POSIXlt`), while the Arrow binding returns a `double` (`POSIXct`) vector.

Reply:
Ah right right. It's a bit more verbose, but what about doing something like:

```
expect_equal(
    df %>%
      arrow_table() %>%
      mutate(
        r_obj_parsed_date = strptime("03-27/2022", format = "%m-%d/%Y"),
        r_obj_parsed_na = strptime("03-27/2022", format = "Y%-%m-%d")) %>%
      collect(),
    df %>%
      arrow_table() %>%
      mutate(
        r_obj_parsed_date = as.POSIXct(strptime("03-27/2022", format = "%m-%d/%Y")),
        r_obj_parsed_na = as.POSIXct(strptime("03-27/2022", format = "Y%-%m-%d")))
    ignore_attr = "tzone"
  )
```

Along with a comment explaining what's going on?

Reply:
Sure, makes a lot of sense (given what `compare_dplyr_binding()` does. I've updated the tests. 1 of 3 uses `record_batch()` and 2 (of 3) use `arrow_table()`.